language/stdlib/modules/Libra.move

```diff
@@ -700,71 +700,159 @@ module Libra {
-        // gets an error for some reason.
+    spec module {
 
-        pragma verify = false;
+        /// Verification is disabled because of an invariant in
+        /// RegisteredCurrencies.move that causes a violated precondition here
+        /// and unprovable invariants here.
+        pragma verify = false, aborts_if_is_partial = true;
     }
 
-    // ## Currency registration
-
     spec module {
-        // Address at which currencies should be registered (mirrors currency_addr)
+        /// Address at which currencies should be registered (mirrors currency_addr)
         define spec_currency_addr(): address { 0xA550C18 }
 
-        // Checks whether currency is registered.
-        // Mirrors is_currency<CoinType> in Move, above.
+        /// Address of Treasury Compliance
+        define spec_treasury_compliance_addr(): address { 0xB1E55ED }
+
+        /// Checks whether currency is registered.
+        /// Mirrors is_currency<CoinType> in Move, above.
         define spec_is_currency<CoinType>(): bool {
             exists<CurrencyInfo<CoinType>>(spec_currency_addr())
         }
     }
 
-    // Sanity check -- after register_currency is called, currency should be registered.
-    spec fun register_currency {
-        // This doesn't verify because:
-        //  1. is_registered assumes the currency is registered at the fixed
-        //     currency_addr()  (0xA550C18).
-        //  2. The address where the CurrencyInfo<CoinType>> is stored is
-        //     determined in Association::initialize()
-        //     (address of AddCurrency privilege) and
-        //     Genesis::initialize_association(association_root_addr).
-        // If the AddCurrency privilege is on an address different from
-        // currency_addr(), the currency will appear not to be registered.
-        // If you change next to "true", prover will report an error.
-        pragma verify = false;
-
-        // SPEC: After register_currency, the currency is an official currency.
-        ensures spec_is_currency<CoinType>();
+    /// ## Management of capabilities
+
+    spec schema OnlyTCHasMintCapabilityInvariant {
+        /// Before a currency is registered, there is no mint capability for that currency.
+        invariant module forall coinType: type, addr: address:
+            !spec_is_currency<coinType>() ==> !exists<MintCapability<coinType>>(addr);
+
+        /// After a currency is registered, only the treasury compliance account
+        /// has the mint capability for that currency.
+        invariant module forall coinType: type, addr: address
+        where spec_is_currency<coinType>():
+            exists<MintCapability<coinType>>(addr)
+                ==> addr == spec_treasury_compliance_addr();
     }
 
-    spec fun is_currency {
-        ensures result == spec_is_currency<CoinType>();
+    spec module {
+        /// The invariant does not hold right now because an account doesn't
```

Comment:
I'm not sure whether I get this. If the requirement to be able to mint is based on the Privilege and not the address, why don't we specify it like this? I remember a discussion a few weeks ago that we wanted to switch from address-based to privilege-based access control.